Check for the presence of the MySQL binaries before calling them

gwolf opened the pull request:
Fixes #388 
Patch to verify for the presence of `mysql` and `mysqldump` (and `gunzip`, as it was in the same command construction) before actually calling it.

Thanks!

gwolf committed:
Check for the presence of the MySQL binaries before calling them

gwolf committed:
Use the proper indentantion and line length for Bee's coding standards

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -348,3 +378,17 @@ function dbq_bee_callback($arguments, $options) {
+/**
+ * Checks whether a command exists in the PATH, and if so, returns the full
+ * path. Returns null if no such command is available.
+ */
```
Please follow the guidance for https://docs.backdropcms.org/doc-standards#functions
adding `@param` and `@return`

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -348,3 +378,17 @@ function dbq_bee_callback($arguments, $options) {
+      return $candidate;
+    }
+  }
+  return null;
```
Please note
https://docs.backdropcms.org/php-standards#naming-conventions

> Constants should always be all-uppercase, with underscores to separate words. This includes pre-defined PHP constants like TRUE, FALSE, and NULL

```suggestion
  return NULL;
```

However, we don't normally use `NULL` as a return value. `FALSE` is simpler as the check for it in this case would be something like:
```php
if (!$cmd_bin) {
```

```suggestion
  return FALSE;
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -119,8 +119,14 @@ function db_export_bee_callback($arguments, $options) {
 
+  $export_command = executable_for('mysqldump');
+  if (is_null($export_command)) {
+    bee_message("The MySQL export command 'mysqldump' cannot be found in " .
+                "this system. Please install it and try again.", 'error');
```
Within Backdrop, there is a function called `t()` that encloses translatable strings.  In bee, because we can't always rely on Backdrop being bootstrapped, we have our own function called `bt()` for translatable strings. Translatable strings are an exception (there are a couple of others) to the 80 character rule and the translatable string should be in a single string.

```suggestion
    bee_message(bt("The MySQL export command 'mysqldump' cannot be found in this system. Please install it and try again."), 'error');
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -119,8 +119,14 @@ function db_export_bee_callback($arguments, $options) {
+  if (is_null($export_command)) {
+    bee_message("The MySQL export command 'mysqldump' cannot be found in " .
+                "this system. Please install it and try again.", 'error');
+    return null;
```
```suggestion
    return FALSE;
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -172,11 +178,23 @@ function db_import_bee_callback($arguments, $options) {
   // Import the database.
+  $mysql_bin = executable_for('mysql');
+  if (is_null($mysql_bin)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
```
Above message about `bt()` applies

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -172,11 +178,23 @@ function db_import_bee_callback($arguments, $options) {
+  if (is_null($mysql_bin)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
+    return null;
```
```suggestion
    return FALSE;
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -172,11 +178,23 @@ function db_import_bee_callback($arguments, $options) {
-    $import_command .= "gunzip -c $filename | ";
+    $gunzip_bin = executable_for('gunzip');
+    if (is_null($gunzip_bin)) {
+      bee_message("The gzip decompressor command 'gunzip' cannot be found " .
+                  "in this system. Please install it and try again.", 'error');
```
Above message about `bt()` applies

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -172,11 +178,23 @@ function db_import_bee_callback($arguments, $options) {
+    if (is_null($gunzip_bin)) {
+      bee_message("The gzip decompressor command 'gunzip' cannot be found " .
+                  "in this system. Please install it and try again.", 'error');
+      return null;
```
```suggestion
      return FALSE;
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -218,7 +236,13 @@ function db_drop_bee_callback($arguments, $options) {
-  $command = 'mysql ' . $connection_string . ' -e "drop database ' . $db_database . '";';
+  $mysql_command = executable_for('mysql');
+  if (is_null($mysql_command)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
```
Above message about `bt()` applies

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -218,7 +236,13 @@ function db_drop_bee_callback($arguments, $options) {
+  if (is_null($mysql_command)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
+    return null;
```
```suggestion
    return FALSE;
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -265,7 +289,13 @@ function sql_bee_callback($arguments, $options) {
-  $command = 'mysql ' . $connection_string;
+  $mysql_command = executable_for('mysql');
+  if (is_null($mysql_command)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
```
Above message about `bt()` applies

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -265,7 +289,13 @@ function sql_bee_callback($arguments, $options) {
+  if (is_null($mysql_command)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
+    return null;
```
```suggestion
    return FALSE;
```

yorkshire-pudding reviewed (changes requested):
Hi @gwolf 

I haven't yet tested this (I don't have any environments without these on to test).  I will run the automated tests again once the changes are made, but might also see if anyone else is able to test the negative conditions.  Most changes are about using `FALSE` instead of `null`, and making the `bee_message` strings translatable. Also a function docblock required.

Thank you

gwolf committed:
Follow Bee coding standards

Use FALSE instead of null.
Write a proper docblock for the introduced function.

gwolf committed:
Use bt() for translatable strings; avoid splitting strings at 80 columns

gwolf commented:
Thanks!
I think I have addressed the needed changes for the coding standards. Sorry, I'm not used to PHP development, and of course, am completely new to this project.

gwolf commented:
Ugh, the codesniffer still does not like my changes. And the details it gives are not really useful (just says "Error:" 😕)

gwolf committed:
Param string should end with a full stop

Aren\'t we a bit over-pedantic? 😉

gwolf commented:
All tests passed! \o/

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
+
+/**
+ * Checks whether a command exists in the PATH, and if so, returns the full
+ * path. Returns FALSE if no such command is available.
```
In line with my suggestion further down about moving this to `includes/filesystem.inc`, it needs a wider explanation of purpose to help other contributors or people writing their own commands.

```suggestion
 * Get the full path for a command executable if it exists in PATH.
 *
 * If your function calls an executable on the system rather than using `bee`,
 * `backdrop` or `php` functions, you can use this function to get the full
 * path of the executable or return FALSE if the executable doesn't exist. The
 * result can then be checked in your calling function, and you can exit with a
 * user friendly error message if the executable doesn't exist in the system.
```

gwolf replied:
Completely makes sense. I have moved the function to the suggested file (and renamed to a more general, project-friendly way).

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
+ * @param string $cmd
+ *   Name of the command to be searched.
+ *
+ * @return string
```
```suggestion
 * @return string|FALSE
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
+ *   Full path to the first occurence of $cmd in the search path.
+ *   Returns FALSE if no matching command is found.
+ */
+function executable_for($cmd) {
```
I've been giving the name of this function a bit of thought, and I've come to the conclusion that this function actually belongs in `includes/filesystem.inc` (`includes` files are all included in `bee.php`) rather than in the `db.bee.inc` file.  Although only db functions currently use this in the core bee software, people writing their own commands or adding a bee command for their module might be grateful for it.  Please could you name the function `bee_get_executable_path` and move to  `includes/filesystem.inc`  - You'll need to update the calls to this function.

Also, can we have full names for variables please.

```suggestion
function bee_get_executable_path($command) {
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
+ *   Returns FALSE if no matching command is found.
+ */
+function executable_for($cmd) {
+  foreach (explode(":", getenv('PATH')) as $dir) {
```
```suggestion
  foreach (explode(":", getenv('PATH')) as $directory) {
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
+ */
+function executable_for($cmd) {
+  foreach (explode(":", getenv('PATH')) as $dir) {
+    $candidate = "$dir/$cmd";
```
```suggestion
    $candidate = "$directory/$command";
```

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
+ * Checks whether a command exists in the PATH, and if so, returns the full
+ * path. Returns FALSE if no such command is available.
+ *
+ * @param string $cmd
```
```suggestion
 * @param string $command
```

yorkshire-pudding reviewed (changes requested):
Sorry @gwolf - a bit more to do.  After thinking about the name of the helper function, I've decided it would be better to go into `includes/filesystem.inc`.  There's also a few tweaks with the function doc bloc and variable names.

gwolf committed:
Further stylistic cleanup

Change variable names to full words instead of abbr
Rename introduced function to a more project-wide descriptive one
Move introduced function to the filesystem includes

yorkshire-pudding commented:
Thanks @gwolf - this is looking good. I tested the negative condition by temporarily changing the defined executable in the `db_export_bee_callback()` function so it couldn't be found.  The function works well.

Thank you for your patience in getting this over the line.
